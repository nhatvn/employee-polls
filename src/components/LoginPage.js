import { useEffect, useState } from "react";
import { connect } from "react-redux";
import { actionLogin } from "../actions/actionLogin";
import userGroup from "../images/user-group.png"
import { useNavigate } from "react-router-dom";
import LoadingBar from "react-redux-loading-bar";
import { handleInitialData, handleInitialDataAfterLogin } from "../actions/shared";
import { objectToArray } from "../utils/helpers";

const LoginPage = ({ dispatch, users }) => {
    useEffect(() => {
        dispatch(handleInitialData());
    }, [dispatch]);

    const navigate = useNavigate();
    const [userName, setUserName] = useState("");
    const [password, setPassword] = useState("");
    const [wrongPassword, setWrongPassword] = useState("");
    const handleUserNameChange = (e) => {
        setUserName(e.target.value);
        let user = users[e.target.value];
        setPassword(user.password);
    };
    const handlePasswordChange = (e) => {
        setPassword(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        let user = users[userName] && users[userName].password === password && users[userName];

        if (user) {
            dispatch(actionLogin(user.id));
            dispatch(handleInitialDataAfterLogin());
            navigate('/');
        }
        else {
            setWrongPassword("Password incorrect.");
        }
    };

    return (
        <div>
            {!users || Object.keys(users).length === 0 ? <LoadingBar /> : (
                <form onSubmit={handleSubmit} data-testid="loginForm">
                    <h3 className="center">Employee Polls</h3>
                    <img src={userGroup} alt="User Avatar" className="user-group" />
                    <h3 className="center">Log In</h3>
                    <div className="mb-10">
                        <label>
                            <div className="center mb-10">User</div>
                            <select onChange={handleUserNameChange}
                                className="form-control"
                                defaultValue=""
                                data-testid="userSelect">
                                <option value="" className="d-none" disabled>User</option>
                                {
                                    objectToArray(users).map((user) => (
                                        <option value={user.id} key={user.id}>{user.id}</option>
                                    ))
                                }
                            </select>
                        </label>
                    </div>
                    <div className="mb-10">
                        <label>
                            <div className="center mb-10">Password</div>
                            <input
                                type="password"
                                name="Password"
                                className="form-control"
                                placeholder="Password"
                                value={password}
                                onChange={handlePasswordChange} />
                        </label>
                    </div>
                    <div className="text-danger" data-testid="wrongPassword">{wrongPassword}</div>
                    <div className="center">
                        <button type="submit" className="btn">Submit</button>
                    </div>
                </form>
            )}
        </div>
    )
};

const mapStateToProps = ({ users }) => ({
    users
});

export default connect(mapStateToProps)(LoginPage);