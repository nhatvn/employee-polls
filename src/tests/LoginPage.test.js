import { fireEvent } from "@testing-library/react";
import LoginPage from "../components/LoginPage";
import { getTestUsers, renderWithProviders } from "./setup-test";

describe('LoginPage', () => {
    it('will match snapshot', () => {
        var component = renderWithProviders(<LoginPage />, {
            preloadedState: {
                users: getTestUsers()
            }
        });
        expect(component).toMatchSnapshot();
    })

    it('will display Password incorrect.', () => {
        var component = renderWithProviders(<LoginPage />, {
            preloadedState: {
                users: getTestUsers()
            }
        });
        var loginForm = component.getByTestId('loginForm');
        fireEvent.submit(loginForm);
        expect(component.getByTestId('wrongPassword').textContent).toEqual('Password incorrect.');
    })

    it('will display login form', async () => {
        var component = renderWithProviders(<LoginPage />, {
            preloadedState: {
                users: getTestUsers()
            }
        });
        var loginForm = component.getByTestId('loginForm');
        expect(loginForm).toBeInTheDocument();
    })
})