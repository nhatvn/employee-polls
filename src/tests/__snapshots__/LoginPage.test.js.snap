// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          style="direction: ltr;"
        >
          <div
            class=""
            style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
          />
          <div
            style="display: table; clear: both;"
          />
        </div>
        <form
          data-testid="loginForm"
        >
          <h3
            class="center"
          >
            Employee Polls
          </h3>
          <img
            alt="User Avatar"
            class="user-group"
            src="user-group.png"
          />
          <h3
            class="center"
          >
            Log In
          </h3>
          <div
            class="mb-10"
          >
            <label>
              <div
                class="center mb-10"
              >
                User
              </div>
              <select
                class="form-control"
                data-testid="userSelect"
              >
                <option
                  class="d-none"
                  disabled=""
                  selected=""
                  value=""
                >
                  User
                </option>
                <option
                  value="sarahedo"
                >
                  sarahedo
                </option>
                <option
                  value="tylermcginnis"
                >
                  tylermcginnis
                </option>
              </select>
            </label>
          </div>
          <div
            class="mb-10"
          >
            <label>
              <div
                class="center mb-10"
              >
                Password
              </div>
              <input
                class="form-control"
                name="Password"
                placeholder="Password"
                type="password"
                value=""
              />
            </label>
          </div>
          <div
            class="text-danger"
            data-testid="wrongPassword"
          />
          <div
            class="center"
          >
            <button
              class="btn"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        style="direction: ltr;"
      >
        <div
          class=""
          style="width: 0%; transition: width 800ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 0;"
        />
        <div
          style="display: table; clear: both;"
        />
      </div>
      <form
        data-testid="loginForm"
      >
        <h3
          class="center"
        >
          Employee Polls
        </h3>
        <img
          alt="User Avatar"
          class="user-group"
          src="user-group.png"
        />
        <h3
          class="center"
        >
          Log In
        </h3>
        <div
          class="mb-10"
        >
          <label>
            <div
              class="center mb-10"
            >
              User
            </div>
            <select
              class="form-control"
              data-testid="userSelect"
            >
              <option
                class="d-none"
                disabled=""
                selected=""
                value=""
              >
                User
              </option>
              <option
                value="sarahedo"
              >
                sarahedo
              </option>
              <option
                value="tylermcginnis"
              >
                tylermcginnis
              </option>
            </select>
          </label>
        </div>
        <div
          class="mb-10"
        >
          <label>
            <div
              class="center mb-10"
            >
              Password
            </div>
            <input
              class="form-control"
              name="Password"
              placeholder="Password"
              type="password"
              value=""
            />
          </label>
        </div>
        <div
          class="text-danger"
          data-testid="wrongPassword"
        />
        <div
          class="center"
        >
          <button
            class="btn"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
